table/block_based_table_reader.cc

```diff
@@ -2737,6 +2766,144 @@ Status BlockBasedTable::Get(const ReadOptions& read_options, const Slice& key,
   return s;
 }
 
+using MultiGetRange = MultiGetContext::Range;
+void BlockBasedTable::MultiGet(const ReadOptions& read_options,
+                               const MultiGetRange* mget_range,
+                               const SliceTransform* prefix_extractor,
+                               bool skip_filters) {
+  const bool no_io = read_options.read_tier == kBlockCacheTier;
+  CachableEntry<FilterBlockReader> filter_entry;
+  bool load_index = false;
+  FilterBlockReader* filter = nullptr;
+  MultiGetRange sst_file_range(*mget_range, mget_range->begin(),
+                               mget_range->end());
+  {
+    if (!skip_filters) {
+      // TODO: Figure out where the stats should go
+      filter_entry = GetFilter(prefix_extractor, /*prefetch_buffer*/ nullptr,
+                               read_options.read_tier == kBlockCacheTier,
+                               nullptr /*get_context*/);
+    }
+    filter = filter_entry.value;
+
+    // First check the full filter
+    // If full filter not useful, Then go into each block
+    FullFilterKeysMayMatch(read_options, filter, &sst_file_range, no_io,
+                           prefix_extractor);
+    for (auto iter = sst_file_range.begin(); iter != sst_file_range.end();
+         ++iter) {
+      load_index = true;
+      break;
+    }
+  }
+  if (load_index) {
+    IndexBlockIter iiter_on_stack;
+    // if prefix_extractor found in block differs from options, disable
+    // BlockPrefixIndex. Only do this check when index_type is kHashSearch.
+    bool need_upper_bound_check = false;
+    if (rep_->index_type == BlockBasedTableOptions::kHashSearch) {
+      need_upper_bound_check = PrefixExtractorChanged(
+          rep_->table_properties.get(), prefix_extractor);
+    }
+    auto iiter = NewIndexIterator(
+        read_options, need_upper_bound_check, &iiter_on_stack,
+        /* index_entry */ nullptr, sst_file_range.begin()->get_context);
+    std::unique_ptr<InternalIteratorBase<BlockHandle>> iiter_unique_ptr;
+    if (iiter != &iiter_on_stack) {
+      iiter_unique_ptr.reset(iiter);
+    }
+
+    for (auto miter = sst_file_range.begin(); miter != sst_file_range.end();
+         ++miter) {
+      Status s;
+      GetContext* get_context = miter->get_context;
+      const Slice& key = miter->ikey;
+      bool matched = false;  // if such user key mathced a key in SST
+      bool done = false;
+      for (iiter->Seek(key); iiter->Valid() && !done; iiter->Next()) {
+        BlockHandle handle = iiter->value();
+
+        bool not_exist_in_filter =
+            filter != nullptr && filter->IsBlockBased() == true &&
+            !filter->KeyMayMatch(ExtractUserKey(key), prefix_extractor,
+                                 handle.offset(), no_io);
```

Comment:
It may be idea to not support block-based filter in this new API. This will simplify the code a little. Just document it in the comments.